backend/src/stwo/prover.rs

```diff
@@ -1,95 +1,315 @@
 }
 
 impl<'a, F: FieldElement, B, MC, C> StwoProver<F, B, MC, C>
 where
-    B: Backend + Send + BackendForChannel<MC>, // Ensure B implements BackendForChannel<MC>
+    B: Backend + Send + BackendForChannel<MC>,
     MC: MerkleChannel + Send,
     C: Channel + Send,
     MC::H: DeserializeOwned + Serialize,
     PowdrComponent<'a, F>: ComponentProver<B>,
 {
     pub fn new(
         analyzed: Arc<Analyzed<F>>,
-        _fixed: Arc<Vec<(String, Vec<F>)>>,
+        fixed: Arc<Vec<(String, VariablySizedColumn<F>)>>,
     ) -> Result<Self, io::Error> {
+        let split: BTreeMap<String, Analyzed<F>> = powdr_backend_utils::split_pil(&analyzed)
+            .into_iter()
+            .collect();
+
         Ok(Self {
             analyzed,
-            _fixed,
-            _proving_key: None,
+            split,
+            fixed,
+            proving_key: StarkProvingKey { preprocessed: None },
             _verifying_key: None,
             _channel_marker: PhantomData,
-            _backend_marker: PhantomData,
             _merkle_channel_marker: PhantomData,
         })
     }
+
+    pub fn set_proving_key(&mut self, rdr: &mut dyn std::io::Read) {
+        let serializable_key: SerializableStarkProvingKey = bincode::deserialize_from(rdr).unwrap();
+        self.proving_key = StarkProvingKey::from(serializable_key);
+    }
+
+    pub fn export_proving_key(
+        &self,
+        writer: &mut dyn std::io::Write,
+    ) -> Result<(), KeyExportError> {
+        let pk = SerializableStarkProvingKey::from(self.proving_key.clone());
+        self.proving_key
+            .preprocessed
+            .as_ref()
+            .ok_or(KeyExportError::NoProvingKey)?;
+        bincode::serialize_into(writer, &pk).unwrap();
+        Ok(())
+    }
+
+    pub fn setup(&mut self) {
+        // machines with varying sizes are not supported yet, and it is checked in backendfactory create function.
+        //TODO: support machines with varying sizes
+        let domain_map: BTreeMap<usize, CircleDomain> = self
+            .analyzed
+            .degrees()
+            .iter()
+            .map(|size| {
+                (
+                    (size.ilog2() as usize),
+                    CanonicCoset::new(size.ilog2()).circle_domain(),
+                )
+            })
+            .collect();
+
+        let preprocessed: BTreeMap<String, TableProvingKeyCollection<B>> = self
+            .split
+            .iter()
+            .filter_map(|(namespace, pil)| {
+                // if we no fixed columns, we don't need to commit to anything, publics is not supported yet.
+                if pil.constant_count() == 0 {
+                    None
+                } else {
+                    let fixed_columns = machine_fixed_columns(&self.fixed, pil);
+
+                    Some((
+                        namespace.to_string(),
+                        //why here it is committed_polys_in_source_order() instead of constant polys?
```

Comment:
Good question actually, I think it does not matter because there is a runtime invariant that all columns inside a namespace have the same degree range, and that we always have at least one witness column. So we find the first witness column and use its degree range, we could do the same with the first constant column.

If this makes sense let's remove this line

Reply:
got it. comment removed